pkg/drain/nodegroup.go

```diff
@@ -127,35 +136,63 @@ func (n *NodeGroupDrainer) Drain() error {
 			}
 
 			if newPendingNodes.Len() == 0 {
-				logger.Success("drained all nodes: %v", drainedNodes.List())
+				waitForAllRoutinesToFinish(ctx, sem, parallelLimit)
+				logger.Success("drained all nodes: %v", mapToList(drainedNodes.Items()))
 				return nil // no new nodes were seen
 			}
 
-			logger.Debug("already drained: %v", drainedNodes.List())
+			logger.Debug("already drained: %v", mapToList(drainedNodes.Items()))
 			logger.Debug("will drain: %v", newPendingNodes.List())
-
 			for i, node := range newPendingNodes.List() {
-				pending, err := n.evictPods(node)
-				if err != nil {
-					logger.Warning("pod eviction error (%q) on node %s", err, node)
-					time.Sleep(retryDelay)
-					continue
-				}
-				logger.Debug("%d pods to be evicted from %s", pending, node)
-				if pending == 0 {
-					drainedNodes.Insert(node)
-				}
+				// scoped values for concurrency
+				i := i
+				node := node
```

Comment:
Rather than doing this, pass it to the go routine.

So alter the go routine to take to parameters, and then pass them in when you call the go routine. This is to ensure proper isolation inside the go routine.